go-controller/pkg/allocator/udn/allocator.go

```diff
@@ -0,0 +1,51 @@
+package udn
+
+import (
+	"fmt"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+type Allocator struct {
+	watchFactory factory.NodeWatchFactory
+	nodeName     string
+}
+
+func New(watchFactory factory.NodeWatchFactory, nodeName string) *Allocator {
+	return &Allocator{
+		watchFactory: watchFactory,
+		nodeName:     nodeName,
+	}
+}
+
+func (a *Allocator) calculateIDsFromNetwork(idName, networkName string, base, limit, cardinality uint) (uint, error) {
+	if cardinality < 1 {
+		return 0, fmt.Errorf("invalid arguments, cardinality '%d' has to be bigger than '1'", cardinality)
+	}
+	if limit <= base {
+		return 0, fmt.Errorf("invalid arguments, limit '%d' has to be bigger than base '%d'", limit, base)
+	}
+	node, err := a.watchFactory.GetNode(a.nodeName)
+	if err != nil {
+		return 0, fmt.Errorf("failed geting node to retrieve network-ids at node '%s' when calculating '%s' for network '%s': %w", a.nodeName, idName, networkName, err)
+	}
+	networkIDByNetworkName, err := util.GetNodeNetworkIDsAnnotationNetworkIDs(node)
+	if err != nil {
+		return 0, fmt.Errorf("failed geting network-ids from node '%s' when calculating '%s' network '%s': %w", a.nodeName, idName, networkName, err)
+	}
+	networkID, ok := networkIDByNetworkName[networkName]
+	if !ok {
+		return 0, fmt.Errorf("missing id for network '%s' when calculating '%s'", networkName, idName)
+	}
+	if networkID < 1 {
+		return 0, fmt.Errorf("invalid arguments, '%s' network-id has to be bigger than '0'", networkName)
+	}
+
+	maxID := base + uint(networkID)*cardinality
+	if maxID >= limit {
+		return 0, fmt.Errorf("out of bounds: calculated max ID '%d' is bigger than limit '%d' for '%s'", maxID, limit, idName)
+	}
+
+	return base + 1 + (cardinality * uint(networkID)) - cardinality, nil
```

Comment:
`base + 1 + (cardinality * uint(networkID)) - cardinality`

can you please explain this approach and why not use the `AllocateNext()` that we seem to use in other places?

Reply:
We decided that instead of using a new bitmap that we have to store at central place like cluster manager we just use the already generated network-id as the "hint" to calculate UDN different IDs.

For example if we have network "blue" with network-id "3" for something with `base` 10 and `cardinality` 2 (two buckets per network), first we have to sum all the backets for that network-id to position at the correct value, that's `cardinality * uint(networkID)`  then we need to calculate the proper base to sum up and get the first number from the two buckets that `10 + 1 - 2` which is `9` so we have a calcualted base of `9` and number of buckets `6` so the first of the two IDs from the bucket of size 2 will be `6` + `9` which is `15`.

In other works this would be the whole distribution for three networks (we skip network id 0 since that's default)
- network-id 1
  - 11
  - 12
- network-id 2
  - 13
  - 14
- network-id 3
  - 15
  - 16

Reply:
Also, u can make new allocator types hold the cardinality as a private property, rather than an argument to this function. Then you don't have to worry about callers of the function accidentally using different cardinalities.

Reply:
> I have no idea what this function is for. It looks like you are trying to make this as a helper to compute values for one of the config values you added in prior commits with a base offset, but the function doesn't make sense to me. Cardinality implies number of numbers in a set, so I interpret this as you want a set of numbers offset by base, but you return a uint.
>

I am going to refactor the the private method to return the final "base" so public functions can increase over it, these way returning just a uint make sense.

I don't want to return a set of IDs since those may need to be loop over again at public mehods to calculate final values, like the case of masquerade.
 
> Using a base of 10, cardinality of 2, and a network id of 1, i get 11. I'm not sure how this is any different form saying base+networkid and returning it.
> 
> If your goal is to include cardinality to make sure you can allocate without exceeding the limit, then you should return the set of things and not just uint.

Validation should be done at the private mehod and we should not return a set of uint so the Public methods are the one iterating it if needed to compose final values so they don't iterate twice.